### PR TITLE
Revise the semantics of facade

### PR DESCRIPTION
**Pull request by mingxwa**
**Major changes**

- Added type `proxy_pointer_constraints`, encapsulating `maximum_size`, `maximum_alignment`, `minimum_copyability`, `minimum_relocatability` and `minimum_destructibility` into a single type.
- Added 3 prototypes of `proxy_pointer_constraints` with popular configurations: `relocatable_pointer_constraints`, `copyable_pointer_constraints` and `trivial_pointer_constraints`.
- Added `concept basic_facade` and `concept facade`.
- Removed class templates `dispatch` and `facade`.
- Added macro `PRO_MAKE_DISPATCH_PACK` as an extension.
- Added `PRO_` prefix to the existing macros to avoid collision.
- Revised the semantics of extension `PRO_DEF_FACADE`, making it easier to use.

**Other changes**
- Replaced the SFINAE guarantee of `basic_facade_traits` with `concept facade`.
- Removed `flattening_traits` in the implementation of `dispatch_traits`.
- Updated unit tests accordingly.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Proxy: Easy Polymorphism in C++
+# Proxy: Polymorphism in C++ Redefined
```
Comment (tian-lt) [inline README.md:1]: nit: Maybe *improved*/*enhanced*😂

File: README.md [hunk 2]
```diff
@@ -60,6 +62,46 @@ pro::proxy<poly::Drawable> CreateRectangleAsDrawable(int width, int height) {
 }
 ```
 
+Here is another demo showing how to define overloads in a dispatch. Note that `.invoke<>` can be ommitted when only 1 dispatch is defined in a facade:
+
+```cpp
+// Abstraction (poly is short for polymorphism)
+namespace poly {
+
+PRO_DEF_MEMBER_DISPATCH(Log, Log,
```
Comment (tian-lt) [inline README.md:71]: nit: use different names to distinguish between dispatch and implementation?

File: proxy.h [hunk 3]
```diff
@@ -16,6 +16,35 @@ namespace pro {
 
 enum class constraint_level { none, nontrivial, nothrow, trivial };
 
+struct proxy_pointer_constraints {
```
Comment (tian-lt) [inline proxy.h:19]: I like it - `proxy_pointer_constraints`.

File: proxy.h [hunk 4]
```diff
@@ -272,13 +294,22 @@ using dependent_t = typename dependent_traits<T, Us...>::type;
 
 }  // namespace details
 
+template <class F>
+concept basic_facade = requires {
+  typename F::dispatch_types;
+  typename std::integral_constant<
```
Comment (tian-lt) [inline proxy.h:300]: Looks like another dance on `std::integral_constant`. Does the standard provide a more specific way to check if a value can be evaluated during compile-time?

**Review by tian-lt — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -102,6 +101,44 @@ int main() {
...
+template <class... Overloads>
+using MyFunction = pro::proxy<poly::MovableCallable<Overloads...>>;
+template <class... Overloads>
+using MyMoveOnlyFunction = pro::proxy<poly::CopyableCallable<Overloads...>>;
+
+int main() {
+  auto f = [](auto&&... v) {
+    printf("f() called. Args: ");
+    ((std::cout << v << ":" << typeid(decltype(v)).name() << ", "), ...);
+    puts("");
```
Comment (tian-lt) [inline README.md:130]: remove `puts("");`? looks like it does nothing.

**Review by tian-lt — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (mingxwa) [reply README.md:130]: It prints an empty line. Otherwise, all the output stays on a same line.

File: README.md (hunk 1 above)
Comment (mingxwa) [reply README.md]: Updated to Next Generation

File: README.md (hunk 2 above)
Comment (mingxwa) [reply README.md]: Removed the second parameter

File: proxy.h (hunk 3 above)
Comment (mingxwa) [reply proxy.h:19]: Glad you like it ^_^

File: proxy.h (hunk 4 above)
Comment (mingxwa) [reply proxy.h:300]: I do not find any

**Pull request merged**